### ISSUE TITLE
Double refund

### ISSUE THREAD

**Issue by LevinMedia** [woocommerce/woocommerce-gateway-stripe#838]
#### Affected ticket(s)

N/A

#### What I expected

When I issue a refund, it should only be refunded once. 

#### What happened instead

The refund doubled up

#### Steps to reproduce the issue

I'm running an WP.com ecomm plan, admining the order through WP-admin - Stripe is in test mode. I'm using card 4242 4242 4242 4242

I placed an order on the front end for one product with multiple items, then visited the edit order screen:

<img width="1560" alt="Screen Shot 2019-04-11 at 1 12 08 PM" src="https://user-images.githubusercontent.com/4500952/55990149-cf3a6200-5c5b-11e9-8c46-9f92c4fa5256.png">

I refunded 1 item:

<img width="1084" alt="Screen Shot 2019-04-11 at 1 16 24 PM" src="https://user-images.githubusercontent.com/4500952/55990337-4c65d700-5c5c-11e9-8220-e0135b2a2629.png">

I clicked "Refund 98 cents via stripe" which resulted in a double refund

<img width="1442" alt="Screen Shot 2019-04-11 at 1 19 07 PM" src="https://user-images.githubusercontent.com/4500952/55990414-7ae3b200-5c5c-11e9-9cf3-60d6c19d1d59.png">

I also see a double refund in my order notes, although the refund ID appears to be the same:

<img width="297" alt="Screen Shot 2019-04-11 at 1 21 23 PM" src="https://user-images.githubusercontent.com/4500952/55990566-d44be100-5c5c-11e9-83a5-09962dda748d.png">

-------------------

- [ ] Issue assigned to next milestone.
- [ ] Issue assigned a priority (will be assessed by maintainers).

**Comment by dechov** [comment]
Can you see if there's a `_stripe_refund_id` in the order meta (e.g. with "Post Meta Inspector" plugin)?

**Comment by MANOJMANI07** [comment]
I have also come across with similar issue. I can see `_stripe_refund_id` in the order meta but it has been added twice and with same refund id. I am also facing an issue with the completed order. On completion of order, I can see two `_stripe_charge_captured`  in the order meta and also `woocommerce_order_status_changed` hook has been called twice.

**Comment by SparkleGear** [comment]
This is still an issue in the current release of WooCommerce 

![image](https://user-images.githubusercontent.com/39808599/73491136-ed548880-437b-11ea-94ac-98e94bdbaf4f.png)

**Comment by SparkleGear** [comment]
Interesting that even though the refund was done through the WooCommerce dashboard, the comment sent with the refund is only on one of the refunds and the second refund says stripe dashboard.

**Comment by dechov** [comment]
Looks like a race condition in which the request [[code](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2c6ef62e9d5a321ef67f91b02b46ba7ada1b607e/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L879)] immediately triggers the webhook event, the handler for which could run the condition on order meta [[code](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2c6ef62e9d5a321ef67f91b02b46ba7ada1b607e/includes/class-wc-stripe-webhook-handler.php#L490)] before it actually gets set upon refund response [[code](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2c6ef62e9d5a321ef67f91b02b46ba7ada1b607e/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L888)]. Seems we need a more robust check for whether the refund has already been initiated in WooCommerce.

**Comment by smilingpeanut** [comment]
+1 to this still occurring. Had it happen for several orders on a site today with WP 5.3.2, WC 3.9.1 and Stripe 4.3.1.

**Comment by dphyled** [comment]
Another report of this in 18057092-hc

![Screenshot](https://d.pr/i/Z0Rkhr+)
Screenshot direct link: https://d.pr/i/Z0Rkhr

Now happens for every refund since update to 4.0

**Comment by dechov** [comment]
Tested with 4.0 and was unable to reproduce the behavior, but was able to after adding a short `sleep` following the request [[code](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2c6ef62e9d5a321ef67f91b02b46ba7ada1b607e/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L879)].

One fix that comes to mind is setting a transitory state during the request and then blocking the webhook handling. For example, `_stripe_refund_id` order meta could be set to some special "pending" value before the request (and unset on error), which can be checked for when the webhook comes in [[here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2c6ef62e9d5a321ef67f91b02b46ba7ada1b607e/includes/class-wc-stripe-webhook-handler.php#L490)].

Another approach would be to just check if that refund ID was already [stored](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2c6ef62e9d5a321ef67f91b02b46ba7ada1b607e/includes/class-wc-stripe-webhook-handler.php#L511), before proceeding with the [rest of the refund processing](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2c6ef62e9d5a321ef67f91b02b46ba7ada1b607e/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L890-L906), but the note would inaccurately say "via Stripe Dashboard".

**Comment by dougaitken** [comment]
https://wordpress.org/support/topic/refunds-are-recorded-twice-in-wp-admin/

**Comment by mouligreenlaw** [comment]
Another instance: 4579950-zen

**Comment by neilmccreadie** [comment]
Another instance in 25341799-hc

**Comment by ArvindKumar08** [comment]
Does anybody know how these issues were resolved? I am facing the same issue with one of our sites.

**Comment by polp6880** [comment]
@ArvindKumar08 my project is actually built on Roots Bedrock/Sage and all I had to do from my end was rebuild and redeploy. The issue hasn't resurfaced since. Definitely a strange one :/

**Comment by glagonikas** [comment]
in case it helps, the duplicates are getting assigned to the root user instead of the person who actually processed the refund.  I assume this will be the case for all other examples, but it goes unnoticed because it's the same user.

<img width="1543" alt="image" src="https://user-images.githubusercontent.com/10516403/159877931-3582e844-2380-4fbf-830e-86356fab3d94.png">

There is a 1s difference between the two and i think this might be because of the stripe webhook (charge.refunded) firing back to us and doubles up to the API request made from the plugin to stripe

<img width="1146" alt="image" src="https://user-images.githubusercontent.com/10516403/159878732-0499c106-7394-45ed-90d8-9589ea0b51ab.png">

This would also explain why one of them is attributed to the root user. The plugin receives the webhook and thinks the order was refunded via the stripe dashboard, but doesn't know who to attribute this to, so it reverts to user id 1 or whoever is the first user of the site.

I would also imagine this only affects partial refunds because on full refunds, it can't really refund any more. I might be wrong on this however.

one solution is to stop those specific webhooks from Stripe since we never process refunds from there, however, I don't think that's the right approach, because we might process the odd one.

Ideally, the plugin should be checking the webhooks and see if the same refund id has already been recorded. if it has, it should ignore it. I can see there is an [attempt to do this](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/6.3.0/includes/class-wc-stripe-webhook-handler.php#L582-L585) but i think it's probably not working correctly?

Also, the order meta have two records with the same refund id
<img width="383" alt="image" src="https://user-images.githubusercontent.com/10516403/159884298-f3ba5b4e-9f31-402a-aaf2-c71c40871a66.png">

**Comment by felipevelzani** [comment]
Reported on 4579950-zen.

**Comment by snandos** [comment]
We have the same problem:
![grafik](https://user-images.githubusercontent.com/42617635/162143506-a845b5ff-a5cf-4ed4-81d1-0ebae5aad7cf.png)

Funny enough, that we have the exact same problem with the PayPal integration too: https://github.com/woocommerce/woocommerce-paypal-payments/issues/522
Might this be a problem with WooCommerce itself?

**Comment by mouligreenlaw** [comment]
Another instance: 5206513-zen

**Comment by Hastibe** [comment]
We're also encountering this issue. I see that this was originally opened several years ago, and is still open--is there any workaround for this, until it is fixed? And/or any timeframe for it to be fixed?

**Comment by samholguin** [comment]
Also encountering this issue. The refund is only processed once on Stripe but showing twice on the website. Once by the user who initiated the refund and the 2nd by the root user.

**Comment by buy** [comment]
Now my baby is almost 1 year old, but this 2018/9 issue is still not resolved ... 😭 
<img width="984" alt="image" src="https://user-images.githubusercontent.com/1431965/192896363-2cab4199-a129-4cdf-8c37-40117b062a97.png">

**Comment by Nfinley** [comment]
Also encountering this issue in the same way @samholguin mentioned. Is there a workaround?

**Comment by snandos** [comment]
> Also encountering this issue in the same way @samholguin mentioned. Is there a workaround?

The workaround we came across was to disable the refund events on the webhook while leaving the other events enabled. This has the downside that any other refund (for example triggered via the Stripe Dashboard) won't be registered by WooCommerce
See https://github.com/woocommerce/woocommerce-paypal-payments/issues/522#issuecomment-1098909297

EDIT: I'm sorry, just saw that @glagonikas already mentioned this approach.

**Comment by xue28** [comment]
6191782-zen

**Comment by stoltzrx** [comment]
6359414 - zen

**Comment by Guilm** [comment]
:)

**Comment by EjayhanFernandes** [comment]
6639400-zen

**Comment by ChrissiePollock** [comment]
#6634648-zen

**Comment by kaushikasomaiya** [comment]
8152570-zen

**Comment by Nemi5150** [comment]
Same problem here. I wish there was a way through the interface to disable the refund webhook, but I don’t see one and I am hesitant to disable in code.

**Comment by Snaacks** [comment]
Zen-8273116

**Comment by wjrosa** [comment]
I tried to reproduce the issue here with no success. I tried using the latest versions of WC and the Stripe extension. With the "legacy checkout experience" enabled and disabled. Webhook calls are received, but the double refund does not happen because of this line (`includes/class-wc-stripe-webhook-handler.php#603`):
```
			// If the refund ID matches, don't continue to prevent double refunding.
			if ( $refund_object->id === $refund_id ) {
				return;
			}
```
It was added on [this PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2499). So maybe some of the users affected above had an older version of the extension. Still, does not explain the latest reports.
